### PR TITLE
OV-1. OV running at testbed mode.

### PR DESCRIPTION
**Pull request by changtengfei**
This PR allows openvisualizer being able gathering serialbytes over opentestbed.

### CODE REVIEW

File: SConstruct [hunk 1]
```diff
@@ -157,6 +157,12 @@ AddOption('--trace',
     action    = 'store_true')
 runnerEnv['TRACEOPT'] = GetOption('traceOpt')
 
+AddOption('--testbed',
+    dest      = 'testbedOpt',
```
Comment (twatteyne) [inline SConstruct]: please rename to `testbed`

File: bin/openVisualizerApp.py [hunk 2]
```diff
@@ -404,6 +412,12 @@ def _addParserArgs(parser):
         action     = 'store',
         help       = 'comma-separated list of IoT-LAB motes (e.g. "wsn430-9,wsn430-34,wsn430-3")'
     )
+    parser.add_argument('-tb', '--testbed',
+        dest       = 'testbedmotes',
+        default    = False,
+        action     = 'store_true',
+        help       = 'comma-separated list of testbed motes'
```
Comment (twatteyne) [inline bin/openVisualizerApp.py]: please test and/or remove this comment.

File: openvisualizer/moteProbe/moteProbe.py [hunk 3]
```diff
@@ -21,18 +21,51 @@
 import time
 import sys
 
+import paho.mqtt.client as mqtt
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py:24]: please add `paho-mqtt` to `requirements.txt`. Please verify name of package.

File: openvisualizer/moteProbe/moteProbe.py [hunk 4]
```diff
@@ -21,18 +21,51 @@
...
 from   pydispatch import dispatcher
 import OpenHdlc
 import openvisualizer.openvisualizer_utils as u
 from   openvisualizer.moteConnector import OpenParser
 from   openvisualizer.moteConnector.SerialTester import SerialTester
 
-#============================ functions =======================================
+#============================ defines =========================================
+
+BROKER_ADDRESS              = "argus.paris.inria.fr"
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: rename all to `OPENTESTBED_*`

File: openvisualizer/moteProbe/moteProbe.py [hunk 5]
```diff
@@ -21,18 +21,51 @@
...
+NOTIFY_SERIALBYTES_TOPIC    = 'opentestbed/deviceType/mote/deviceId/+/notif/frommoteserialbytes'
+TOTAL_NUMBER_MOTES_TESTBED  = 76
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: change logic: in `findSerialPorts`, issue a `opentestbed/deviceType/box/deviceId/all/cmd/status` command and wait 10s for all responses.

File: openvisualizer/moteProbe/moteProbe.py [hunk 6]
```diff
@@ -21,18 +21,51 @@
...
+DISCOVER_MOTES_TIMEOUT      = 30
 
 BAUDRATE_LOCAL_BOARD  = 115200
 BAUDRATE_IOTLAB       = 500000
 
-def findSerialPorts(isIotMotes=False):
+#============================ variables =======================================
+
+
+testbedmotes = []
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: `opentestedmotes`

File: openvisualizer/moteProbe/moteProbe.py [hunk 7]
```diff
@@ -21,18 +21,51 @@
...
+discovermotes_queue = Queue.Queue()
+
+#============================ functions =======================================
+
+def on_connect(client, userdata, flags, rc):
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: rename with `opentestbed_` prefix

File: openvisualizer/moteProbe/moteProbe.py [hunk 8]
```diff
@@ -21,18 +21,51 @@
...
+
+    print "connect to :", NOTIFY_SERIALBYTES_TOPIC
+    client.subscribe(NOTIFY_SERIALBYTES_TOPIC)
+
+
+def on_message(client, userdata, message):
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: rename with `opentestbed_` prefix

File: openvisualizer/moteProbe/moteProbe.py [hunk 9]
```diff
@@ -21,18 +21,51 @@
...
+
+    # get the motes list from payload
+    
+    testbedmotes.append(message.topic.split('/')[4])
+    
+    if len(testbedmotes) == TOTAL_NUMBER_MOTES_TESTBED:
+        discovermotes_queue.put('all motes responsed')
+
+
+def findSerialPorts(isIotMotes=False, isTestbedMotes=False):
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: `isOpentestbedMotes`

Comment (twatteyne) [reply openvisualizer/moteProbe/moteProbe.py]: rework the `findSerialPorts` helper function by splitting it in 3 functions (serial, IoT-lab, OpenTestbed).

File: openvisualizer/moteProbe/moteProbe.py [hunk 10]
```diff
@@ -96,27 +147,37 @@ class moteProbe(threading.Thread):
     MODE_SERIAL    = 'serial'
     MODE_EMULATED  = 'emulated'
     MODE_IOTLAB    = 'IoT-LAB'
+    MODE_TESTBED   = 'opentestbed'
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py:158]: `MODE_OPENTESTBED`

Comment (twatteyne) [reply openvisualizer/moteProbe/moteProbe.py:158]: always all lowercase: `opentestbed`

File: openvisualizer/moteProbe/moteProbe.py [hunk 11]
```diff
@@ -131,6 +192,9 @@ def __init__(self,serialport=None,emulatedMote=None,iotlabmote=None):
         elif self.mode==self.MODE_IOTLAB:
             self.iotlabmote       = iotlabmote
             self.portname         = 'IoT-LAB{0}'.format(iotlabmote)
+        elif self.mode==self.MODE_TESTBED:
+            self.testbedmote      = testbedmote
+            self.portname         = 'opentestbed{0}'.format(testbedmote)
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: rename `testbedmote` to `testbedmote_eui64`

Comment (twatteyne) [reply openvisualizer/moteProbe/moteProbe.py]: `self.portname         = 'opentestbed{0}'.format(testbedmote)`
to
self.portname         = 'opentestbed_{0}'.format(testbedmote)

File: openvisualizer/moteProbe/moteProbe.py [hunk 12]
```diff
@@ -190,6 +266,9 @@ def run(self):
                 elif self.mode==self.MODE_IOTLAB:
                     self.serial = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
                     self.serial.connect((self.iotlabmote,20000))
+                elif self.mode==self.MODE_TESTBED:
+                    # subscribe to topic: opentestbed/deviceType/mote/deviceId/00-12-4b-00-14-b5-b6-49/notif/frommoteserialbytes
+                    self.serial = self.serialbytes_queue
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: rename `serial` to `mqtt_seriaqueue`

File: openvisualizer/moteProbe/moteProbe.py [hunk 13]
```diff
@@ -303,3 +385,18 @@ def _bufferDataToSend(self,data):
...
             self.outputBuf += [hdlcData]
+            
+    #==== mqtt callback functions
+    
+    def _on_mqtt_connect(self, client, userdata, flags, rc):
+    
+        print "connect to :", self.mqttclient_topic_format.format(self.testbedmote)
+        
+        client.subscribe(self.mqttclient_topic_format.format(self.testbedmote))
+        # print "subscribe at {0}".format(self.mqttclient_topic_format)
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: remove

File: openvisualizer/moteProbe/moteProbe.py [hunk 14]
```diff
@@ -303,3 +385,18 @@ def _bufferDataToSend(self,data):
...
+        
+        client.loop_start()
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: why two `loop_start()` calls?

**Review by twatteyne — CHANGES_REQUESTED** [review]
looks great, some little suggestions

File: openvisualizer/moteProbe/moteProbe.py [hunk 15]
```diff
@@ -21,18 +21,58 @@
...
-def findSerialPorts(isIotMotes=False):
+#============================ variables =======================================
+
+opentestbed_motelist = []
+
+#============================ functions =======================================
+
+def on_connect(client, userdata, flags, rc):
+    
+    print "connect to :", OPENTESTBED_REPS_STATUS_TOPIC
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: connected to

File: openvisualizer/moteProbe/moteProbe.py [hunk 16]
```diff
@@ -21,18 +21,58 @@
...
+    client.subscribe(OPENTESTBED_REPS_STATUS_TOPIC)
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: REPS -> RESP

File: openvisualizer/moteProbe/moteProbe.py [hunk 17]
```diff
@@ -21,18 +21,58 @@
...
 import OpenHdlc
 import openvisualizer.openvisualizer_utils as u
 from   openvisualizer.moteConnector import OpenParser
 from   openvisualizer.moteConnector.SerialTester import SerialTester
 
-#============================ functions =======================================
+#============================ defines =========================================
+
+OPENTESTBED_BROKER_ADDRESS          = "argus.paris.inria.fr"
+OPENTESTBED_CMD_STATUS_TOPIC        = 'opentestbed/deviceType/box/deviceId/all/cmd/status'
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: use directly in the code, no variable needed

File: openvisualizer/moteProbe/moteProbe.py [hunk 18]
```diff
@@ -21,18 +21,58 @@
...
+        'token':       123,
+    }
+    # publish the cmd message
+    client.publish(
+        topic   = OPENTESTBED_CMD_STATUS_TOPIC,
+        payload = json.dumps(payload_status),
+    )
+
+
+def on_message(client, userdata, message):
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: needs to be part of a class

File: openvisualizer/moteProbe/moteProbe.py [hunk 19]
```diff
@@ -151,6 +223,18 @@ def __init__(self,serialport=None,emulatedMote=None,iotlabmote=None):
         # flag to permit exit from read loop
         self.goOn                 = True
         
+        if self.mode == self.MODE_TESTBED:
+            # initialize variable for testbedmote
+            self.mqttclient_topic_format = 'opentestbed/deviceType/mote/deviceId/{0}/notif/frommoteserialbytes'
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: don;'t use a variable

File: openvisualizer/moteProbe/moteProbe.py [hunk 20]
```diff
@@ -303,3 +393,13 @@ def _bufferDataToSend(self,data):
...
+            
+    #==== mqtt callback functions
+    
+    def _on_mqtt_connect(self, client, userdata, flags, rc):
+        
+        client.subscribe(self.mqttclient_topic_format.format(self.testbedmote_eui64))
+        
+    def _on_mqtt_message(self, client, userdata, message):
+    
+        self.serialbytes_queue.put(json.loads(message.payload)['serialbytes'])
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: add a try/except around this. Print/log a big error, but don't crash

File: openvisualizer/moteProbe/moteProbe.py [hunk 21]
```diff
@@ -150,6 +99,59 @@ def findSerialPorts(isIotMotes=False, isTestbedMotes=False):
 
 #============================ class ===========================================
 
+class findOpenTestbedMotes(object):
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: rename to `OpentestbedMoteFinder` (this is a class, not a function)

File: openvisualizer/moteProbe/moteProbe.py [hunk 22]
```diff
@@ -150,6 +99,59 @@ def findSerialPorts(isIotMotes=False, isTestbedMotes=False):
...
+        time.sleep(self.OPENTESTBED_RESP_STATUS_TIMEOUT)
+        
+        # close the client and return the motes list
+        self.mqtt_client.loop_stop()
+        
+        print "{0} motes are found".format(len(self.opentestbed_motelist))
+        
+    def get_opentestbed_motelist(self):
+        
+        return self.opentestbed_motelist
```
Comment (twatteyne) [inline openvisualizer/moteProbe/moteProbe.py]: race condition. How do you ensure you call `get_opentestbed_motelist()` after the motes have been discovered?

**Review by twatteyne — CHANGES_REQUESTED** [review]

File: openvisualizer/moteProbe/moteProbe.py (hunk 22 above)
Comment (changtengfei) [reply openvisualizer/moteProbe/moteProbe.py]: @twatteyne In the _init_ function, it already wait for 10 seconds. That will make sure the messages from all otboxes are received.

**Comment by twatteyne** [comment]
:+1:

**Pull request merged**